common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -117,76 +119,438 @@ public Collation(
+        UTF8_BINARY, ICU, INDETERMINATE
+      }
+
+      protected enum CaseSensitivity {
+        CS, CI
+      }
+
+      protected enum AccentSensitivity {
+        AS, AI
+      }
+
+      protected enum CaseConversion {
+        UNSPECIFIED, LCASE, UCASE
+      }
+
+      protected static final int IMPLEMENTATION_PROVIDER_OFFSET = 29;
+      protected static final int IMPLEMENTATION_PROVIDER_MASK = 0b11;
+      protected static final int CASE_SENSITIVITY_OFFSET = 28;
+      protected static final int CASE_SENSITIVITY_MASK = 0b1;
+      protected static final int ACCENT_SENSITIVITY_OFFSET = 27;
+      protected static final int ACCENT_SENSITIVITY_MASK = 0b1;
+      protected static final int CASE_CONVERSION_OFFSET = 21;
+      protected static final int CASE_CONVERSION_MASK = 0b11;
+      protected static final int LOCALE_OFFSET = 0;
+      protected static final int LOCALE_MASK = 0x0FFF;
+
+      protected static final int INDETERMINATE_COLLATION_ID =
+        ImplementationProvider.INDETERMINATE.ordinal() << IMPLEMENTATION_PROVIDER_OFFSET;
+
+      protected final CaseSensitivity caseSensitivity;
+      protected final AccentSensitivity accentSensitivity;
+      protected final CaseConversion caseConversion;
+      protected final String locale;
+      protected final int collationId;
+
+      protected CollationSpec(
+          String locale,
+          CaseSensitivity caseSensitivity,
+          AccentSensitivity accentSensitivity,
+          CaseConversion caseConversion) {
+        this.locale = locale;
+        this.caseSensitivity = caseSensitivity;
+        this.accentSensitivity = accentSensitivity;
+        this.caseConversion = caseConversion;
+        this.collationId = getCollationId();
+      }
+
+      private static final Map<Integer, Collation> collationMap = new ConcurrentHashMap<>();
+
+      public static Collation fetchCollation(int collationId) throws SparkException {
+        if (collationMap.containsKey(collationId)) {
+          return collationMap.get(collationId);
+        } else {
+          CollationSpec spec;
+          int implementationProviderOrdinal =
+            (collationId >> IMPLEMENTATION_PROVIDER_OFFSET) & IMPLEMENTATION_PROVIDER_MASK;
+          if (implementationProviderOrdinal >= ImplementationProvider.values().length) {
+            throw SparkException.internalError("Invalid implementation provider");
+          } else {
+            ImplementationProvider implementationProvider = ImplementationProvider.values()[
+              implementationProviderOrdinal];
+            if (implementationProvider == ImplementationProvider.UTF8_BINARY) {
+              spec = CollationSpecUTF8Binary.fromCollationId(collationId);
+            } else if (implementationProvider == ImplementationProvider.ICU) {
+              spec = CollationSpecICU.fromCollationId(collationId);
+            } else {
+              throw SparkException.internalError("Cannot instantiate indeterminate collation");
```

Comment:
Would this be hit if user says `collate(StringType(INDETERMINATE))`? If so than this can't be internal error.

Reply:
I don't think user can specify `INDETERMINATE` collation name explicitly. This function can only be called internally with `collationId` integer parameter which is not exposed to user; clients of this function are other functions which parse string name first to integer collation ID and then pass it around.

Reply:
Actually I'm not sure about this, indeterminate collation id may be propagate to places like https://github.com/apache/spark/blob/b20356ef55a230fe00d0463757fb327aefe5ccff/sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala#L38

Reply:
I believe for now it is not possible to create `INDETERMINATE` collation. What we do need is to be able to alias the values with this collation in select queries (this is a print of result of for example concat function). This is something tricky which I would say we could deal with after this PR, as we need to allow printing, but not storing of the INDETERMINATE collation, but for now I believe it is sufficient to just block users from creating INDETERMINATE collation explicitly with COLLATE in SQL. Same blocking should apply to pyspark and dataframe API.

Reply:
Outdated. We don't throw internal errors anymore.